`Bug` [Samsung_HVAC]: ESPHome 2026.1.0 update error

### The problem

216
Hello, after updating ESP Home to version 2026.1.0, a compilation error is occurring when updating devices with Samsung_HVAC.

### ESPHome Version

2026.1.0

### Samsung HVAC Component Version

main

### Home Assistant Version

2026.1.2

### AC Type

NASA

### Indoor AC Unit Model

AR18CVFAAWKNAZ

### Outdoor AC Unit Model

AR18CVFAAWKXAZ

### ESP Device Model

M5STACK ATOM Lite

### Connection Points

F1/F2

### ESPHome Logs

Processing /config/esphome/ac-escritorio-esp.yaml
------------------------------------------------------------

INFO ESPHome 2026.1.0
INFO Reading configuration /config/esphome/ac-escritorio-esp.yaml...
INFO Generating C++ source...
INFO Setting CONFIG_LWIP_MAX_SOCKETS to 14 (registered: api=4, captive_portal=4, mdns=2, ota=1, web_server=3)
INFO Compiling app... Build path: /data/build/samsung-ac-escritorio
Processing samsung-ac-escritorio (board: m5stack-atom; framework: espidf; platform: https://github.com/pioarduino/platform-espressif32/releases/download/55.03.35/platform-espressif32.zip)
--------------------------------------------------------------------------------
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
 - contrib-piohome @ 3.4.4 
 - framework-espidf @ 3.50502.0 (5.5.2) 
 - tool-cmake @ 4.0.3 
 - tool-esp-rom-elfs @ 2024.10.11 
 - tool-esptoolpy @ 5.1.0 
 - tool-mklittlefs @ 3.2.0 
 - tool-ninja @ 1.13.1 
 - tool-scons @ 4.40801.0 (4.8.1) 
 - toolchain-xtensa-esp-elf @ 14.2.0+20251107
Reading CMake configuration...
Generating assembly for certificate bundle...
Dependency Graph
|-- noise-c @ 0.1.10
|-- ArduinoJson @ 7.4.2
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/api/api_connection.cpp.o
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/api/api_frame_helper.cpp.o
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/api/api_frame_helper_noise.cpp.o
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/api/api_pb2.cpp.o
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/api/api_pb2_service.cpp.o
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/api/api_server.cpp.o
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/api/list_entities.cpp.o
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/api/proto.cpp.o
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/api/subscribe_state.cpp.o
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/captive_portal/captive_portal.cpp.o
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/captive_portal/dns_server_esp32_idf.cpp.o
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/climate/climate.cpp.o
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/climate/climate_mode.cpp.o
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/climate/climate_traits.cpp.o
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/esp32/core.cpp.o
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/esp32/gpio.cpp.o
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/esp32/helpers.cpp.o
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/esp32/preferences.cpp.o
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/esphome/ota/ota_esphome.cpp.o
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/json/json_util.cpp.o
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/logger/logger.cpp.o
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/logger/logger_esp32.cpp.o
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/logger/task_log_buffer_esp32.cpp.o
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/md5/md5.cpp.o
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/mdns/mdns_component.cpp.o
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/mdns/mdns_esp32.cpp.o
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/network/util.cpp.o
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/number/automation.cpp.o
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/number/number.cpp.o
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/number/number_call.cpp.o
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/number/number_traits.cpp.o
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/ota/ota_backend.cpp.o
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/ota/ota_backend_esp_idf.cpp.o
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/ota/ota_backend_host.cpp.o
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/safe_mode/safe_mode.cpp.o
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/samsung_ac/conversions.cpp.o
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/samsung_ac/debug_mqtt.cpp.o
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/samsung_ac/log.cpp.o
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/samsung_ac/protocol.cpp.o
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/samsung_ac/protocol_nasa.cpp.o
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/samsung_ac/protocol_non_nasa.cpp.o
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/samsung_ac/samsung_ac.cpp.o
Compiling .pioenvs/samsung-ac-escritorio/src/esphome/components/samsung_ac/samsung_ac_device.cpp.o
src/esphome/components/samsung_ac/samsung_ac_device.cpp: In member function 'virtual void esphome::samsung_ac::Samsung_AC_Climate::control(const esphome::climate::ClimateCall&)':
src/esphome/components/samsung_ac/samsung_ac_device.cpp:118:56: error: cannot convert 'esphome::StringRef' to 'const char*' in initialization
  118 |       const char *custom_fan = call.get_custom_fan_mode();
      |                                ~~~~~~~~~~~~~~~~~~~~~~~~^~
      |                                                        |
      |                                                        esphome::StringRef
src/esphome/components/samsung_ac/samsung_ac_device.cpp:139:57: error: cannot convert 'esphome::StringRef' to 'const char*' in initialization
  139 |       const char *custom_preset = call.get_custom_preset();
      |                                   ~~~~~~~~~~~~~~~~~~~~~~^~
      |                                                         |
      |                                                         esphome::StringRef
*** [.pioenvs/samsung-ac-escritorio/src/esphome/components/samsung_ac/samsung_ac_device.cpp.o] Error 1
========================= [FAILED] Took 185.41 seconds =========================

### Samsung AC Logs

_No response_

### YAML Configuration

```yaml
# This example provides a basic configuration and description based on the M5STACK components
# For further configuration you should visit the ESPHome docs https://esphome.io
# Every line starting with a # can be removed.

## BASIC CONFIGURATION
esphome:
  name: samsung-ac-escritorio
  friendly_name: AC_Escritorio_ESP

# Select chipset and board configuration - in this case M5STACK
esp32:
  board: m5stack-atom
  framework:
    type: esp-idf

# [Optional] - Create web interface where you can control the AC from your web browser (without Home Assistant)
web_server:
  port: 80

# Enable logging
logger:
  logs:
    component: ERROR # Remove the "Your component takes to long to respond warning"

# Enable native API for communication with Home Assistant, this can be removed if you do not plan to use it with Home Assistant
api:
  encryption:
    key: "###################################"

# Enable OTA firmware update, you will be able to upload the firmware over WiFi without connecting the device by cables
ota:
  - platform: esphome
    password: "###########################"

# Setup your WiFi password
wifi:
  ssid: !secret wifi_ssid # !!!CHANGE ME!!!
  password: !secret wifi_password # !!!CHANGE ME!!!
  enable_btm: true  # 802.11v BSS Transition Management
  enable_rrm: true  # 802.11k Radio Resource Management
  
  ap:
    ssid: "Ac-Samsung-Escritorio"
    password: "##########################"

captive_portal:

# Specify pins used by the board to comunicate with RS485 board - in this case M5STACK
uart:
  tx_pin: GPIO19   # Use GPIO26 for ATOM Tail485 
  rx_pin: GPIO22   # Use GPIO32 for ATOM Tail485
  # For older devices it could also be 2400. But you should start with 9600.
  baud_rate: 9600
  parity: EVEN

## SAMSUNG AC CONFIGURATION

# Import custom component from GitHub
external_components:
  - source: github://omerfaruk-aran/esphome_samsung_hvac_bus@main # Use @stable if you want a more stable experience, but note a lot of this example code breaks.
    components: [samsung_ac]

# Configuration of AC component
samsung_ac:

  # For NonNASA devices the following option can be enabled to prevent the device from sleeping when idle. This allows
  # values like internal and external temperature to continue to be tracked when the device isn't in use.
  non_nasa_keepalive: true
  
  # When enabled (set to true), this option will log the messages associated with undefined codes on the device. This is useful for debugging and identifying any unexpected or unknown codes that the device may receive during operation.
  debug_log_undefined_messages: false
  
  # When enabled (set to true), this option logs messages associated with defined codes on the device. This helps in monitoring the behavior of the device by recording the activity related to known, expected codes.
  debug_log_messages: false

  # Capabilities configure the features that all devices of your AC system have (all parts of this section are optional). 
  # All capabilities are off by default, you need to enable only those your devices have.
  # You can override or configure them also on a per-device basis (look below for that).
  capabilities: 
    vertical_swing: true
    horizontal_swing: true
    # Presets define special AC modes like Windfree, Eco, and so on. 
    # The following modes are available: sleep, quiet, fast, longreach, windfree, eco.
    presets: 
      # Short version - a quick and simple way to enable presets.
      quiet: true
      sleep: true
      windfree: true
      eco: true
      # Long version - allows customization, including localized names and control over enabling the presets.
      fast:
        name: "Fast cooling"
        enabled: true

  devices:
    # Skip everything below on the first run! Wait a minute, watch your ESPHome logs until you see the "Discovered devices:" section and you see some addresses:
    # For NASA devices it looks like this (with 1 outdoor device and 4 indoor devices): 
    #   Outdoor: 10.00.00
    #   Indoor:  20.00.00, 20.00.01, 20.00.02, 20.00.03
    # For NonNASA devices it looks like this (with 2 indoor devices): 
    #   Indoor:  00, 01
    # You need these addresses for the the sections below. If you see only a "-" then you should visit the troubleshooting section in the readme. 

    # Repeat everything below for each indoor device address you see in your logs
    - address: "20.00.00" # Indoor device address 
      # Each property below is optional (climate, room_temperature etc.) - you can delete those which you dont need.
      # For the names we suggest to choose a combination of room name and the thing it controls. 

      # Configures/overrides the capabilites for this devices. 
      # Look above for all options.
      capabilities: 
        vertical_swing: true
        horizontal_swing: true
        # Presets define special AC modes like Windfree, Eco, and so on. 
        # The following modes are available: sleep, quiet, fast, longreach, windfree, eco.
        presets: 
          # Short version - a quick and simple way to enable presets.
          quiet:
            name: "Silencioso"
            enabled: true
          sleep:
            name: "Dormir"
            enabled: true
          windfree:
            name: "Windfree"
            enabled: true
          eco:
            name: "Economico"
            enabled: true
          fast:
            name: "Turbo"
            enabled: true

      # Creates climate control in Home Assistant. A climate control combines multiple of the controls below (like temperature, mode etc.)
      climate:
        name: "Escritorio climate"

      # The controls directly below are all included in the climate control. Its adviced to only add the climate control and skip the extra controls.
      room_temperature:
        name: "Escritorio temperature"
      target_temperature:
        name: "Escritorio target temperature"
      power:
        name: "Escritorio power"
      mode:
        name: "Escritorio mode"

      #Add automatic_cleaning mode for Samsung AC NASA
            #- Added the "automatic_cleaning" feature to the Samsung AC integration.
            #- This mode ensures hygienic drying of all moisture in the indoor unit after the cooling process is completed.
      automatic_cleaning:
        name: "Escritorio automatic clean"

      # If your AC sits near or inside the ceiling, the reported room temperature is often a little bit heigher then whats 
      # measured below. This property can be used to correct that value.
      #room_temperature_offset: -1.4
      room_temperature_offset: 0

      # Only supported on NASA devices
      room_humidity:
        name: "Escritorio humidity"
      
    - address: "10.00.00" # Outdoor device address as the following components are dependent on an outdoor unit.
      # This sensor captures and monitors specific error codes returned by the HVAC system.
      # When an error occurs, the sensor detects the error code and updates its value accordingly.
      # Additionally, by using the blueprint available at https://github.com/omerfaruk-aran/esphome_samsung_ac_blueprint,
      # you can automatically send detailed error messages to your mobile devices based on the captured error codes.
      error_code:
        name: error_code
        
      # This sensor measures the instantaneous power consumption of the outdoor unit in Watts.
      # The captured value represents the current power draw of the outdoor HVAC components, helping track energy usage patterns.
      outdoor_instantaneous_power:
        name: "Outdoor Instantaneous Power"

      # This sensor records the cumulative energy consumption of the outdoor unit in kWh.
      # It calculates the total energy consumed over time, allowing users to monitor and analyze energy efficiency.
      outdoor_cumulative_energy:
        name: "Outdoor Cumulative Energy"

      # This sensor measures the current drawn by the outdoor unit in Amperes.
      # Monitoring current values helps identify electrical irregularities and ensure safe power levels in the system.
      outdoor_current:
        name: "Outdoor Current"

      # This sensor tracks the voltage supplied to the outdoor unit in Volts.
      # Consistent voltage readings indicate stable power delivery, while deviations can help detect electrical issues in the system.
      outdoor_voltage:
        name: "Outdoor Voltage"

      # Only supported on NASA based heatpumps
      outdoor_temperature: # Should be used with outdoor device address
        name: "Outdoor temperature"
```

### Additional Details

_No response_

--- a/components/samsung_ac/samsung_ac_device.cpp
+++ b/components/samsung_ac/samsung_ac_device.cpp
@@ -115,16 +115,21 @@ namespace esphome
       }
 
       auto fanmodeOpt = call.get_fan_mode();
-      const char *custom_fan = call.get_custom_fan_mode();
-
+      // const char *custom_fan = call.get_custom_fan_mode();
+      auto custom_fan = call.get_custom_fan_mode();
+      
       if (fanmodeOpt.has_value())
       {
         request.fan_mode = climatefanmode_to_fanmode(fanmodeOpt.value());
       }
-      else if (custom_fan != nullptr && custom_fan[0] != '\0')
-      {
-        request.fan_mode = customfanmode_to_fanmode(custom_fan);
+      // else if (custom_fan != nullptr && custom_fan[0] != '\0')
+      // {
+      //   request.fan_mode = customfanmode_to_fanmode(custom_fan);
+      // }
+      else if (!custom_fan.empty()) {
+        request.fan_mode = customfanmode_to_fanmode(custom_fan.c_str());
       }
+     
       else if (mode_changed)
       {
         FanMode auto_fan = climatefanmode_to_fanmode(climate::CLIMATE_FAN_AUTO);
@@ -136,12 +141,17 @@ namespace esphome
         set_alt_mode_by_name(request, preset_to_altmodename(presetOpt.value()));
       }
 
-      const char *custom_preset = call.get_custom_preset();
-      if (custom_preset != nullptr && custom_preset[0] != '\0')
-      {
-        set_alt_mode_by_name(request, AltModeName(custom_preset));
-      }
+      // const char *custom_preset = call.get_custom_preset();
+      auto custom_preset = call.get_custom_preset();
 
+      // if (custom_preset != nullptr && custom_preset[0] != '\0')
+      // {
+      //   set_alt_mode_by_name(request, AltModeName(custom_preset));
+      // }
+      if (!custom_preset.empty()) {
+        set_alt_mode_by_name(request, AltModeName(custom_preset.c_str()));
+      }
+      
       auto swingModeOpt = call.get_swing_mode();
       if (swingModeOpt.has_value())
       {
@@ -201,4 +211,5 @@ namespace esphome
 
   } // namespace samsung_ac
 } // namespace esphome
+
 

--- a/components/samsung_ac/samsung_ac_device.cpp
+++ b/components/samsung_ac/samsung_ac_device.cpp
@@ -126,7 +126,8 @@ namespace esphome
       // {
       //   request.fan_mode = customfanmode_to_fanmode(custom_fan);
       // }
-      else if (!custom_fan.empty()) {
+      else if (!custom_fan.empty()) 
+      {
         request.fan_mode = customfanmode_to_fanmode(custom_fan.c_str());
       }
      
@@ -211,5 +212,6 @@ namespace esphome
 
   } // namespace samsung_ac
 } // namespace esphome
+
 
 

--- a/components/samsung_ac/samsung_ac_device.cpp
+++ b/components/samsung_ac/samsung_ac_device.cpp
@@ -115,17 +115,12 @@ namespace esphome
       }
 
       auto fanmodeOpt = call.get_fan_mode();
-      // const char *custom_fan = call.get_custom_fan_mode();
       auto custom_fan = call.get_custom_fan_mode();
       
       if (fanmodeOpt.has_value())
       {
         request.fan_mode = climatefanmode_to_fanmode(fanmodeOpt.value());
       }
-      // else if (custom_fan != nullptr && custom_fan[0] != '\0')
-      // {
-      //   request.fan_mode = customfanmode_to_fanmode(custom_fan);
-      // }
       else if (!custom_fan.empty()) 
       {
         request.fan_mode = customfanmode_to_fanmode(custom_fan.c_str());
@@ -141,14 +136,7 @@ namespace esphome
       {
         set_alt_mode_by_name(request, preset_to_altmodename(presetOpt.value()));
       }
-
-      // const char *custom_preset = call.get_custom_preset();
       auto custom_preset = call.get_custom_preset();
-
-      // if (custom_preset != nullptr && custom_preset[0] != '\0')
-      // {
-      //   set_alt_mode_by_name(request, AltModeName(custom_preset));
-      // }
       if (!custom_preset.empty()) {
         set_alt_mode_by_name(request, AltModeName(custom_preset.c_str()));
       }
@@ -212,6 +200,7 @@ namespace esphome
 
   } // namespace samsung_ac
 } // namespace esphome
+
 
 
 

